File: zeroinstall/injector/requirements.py
"""
Holds information about what the user asked for (which program, version constraints, etc).
"""

# Copyright (C) 2012, Thomas Leonard
# See the README file for details, or visit http://0install.net.

from zeroinstall import SafeException

class Requirements(object):
	"""
	Holds information about what the user asked for (which program, version constraints, etc).
	"""

	# Note: apps.py serialises our __slots__
	__slots__ = [
		'interface_uri',
		'command',
		'source',
		'extra_restrictions',		# {str: str} (iface -> range)
		'os', 'cpu',
		'message',
	]

	def __init__(self, interface_uri):
		"""@type interface_uri: str"""
		self.interface_uri = interface_uri
		self.command = 'run'
		self.source = False
		self.os = self.cpu = None
		self.message = None
		self.extra_restrictions = {}

	def parse_options(self, options):
		self.extra_restrictions = self._handle_restrictions(options)

		for uri, expr in self.extra_restrictions.items():
			if not expr:
				raise SafeException("Missing version expression for {uri}".format(uri = uri))

		self.source = bool(options.source)
		self.message = options.message

		self.cpu = options.cpu
		self.os = options.os

		# (None becomes 'run', while '' becomes None)
		if options.command is None:
			if self.source:
				self.command = 'compile'
			else:
				self.command = 'run'
		else:
			self.command = options.command or None

	def parse_update_options(self, options):
		"""Update the settings based on the options (used for "0install update APP").
		@return: whether any settings were changed
		@rtype: bool
		@since: 1.9"""
		restriction_updates = self._handle_restrictions(options)

		changed = False
		for key in ['message', 'cpu', 'os', 'command']:
			value = getattr(options, key)
			if value is not None:
				changed = changed or value != getattr(self, key)
				setattr(self, key, value)

		for uri, expr in restriction_updates.items():
			old = self.extra_restrictions.get(uri, None)
			changed = expr != old
			if expr:
				self.extra_restrictions[uri] = expr
			elif uri in self.extra_restrictions:
				del self.extra_restrictions[uri]

		if options.source and not self.source:
			# (partly because it doesn't make much sense, and partly because you
			# can't undo it, as there's no --not-source option)
			raise SafeException("Can't update from binary to source type!")
		return changed

	def get_as_options(self):
		"""@rtype: [str]"""
		gui_args = []
		if self.extra_restrictions:
			# Currently, we only handle the case of restrictions on the root
			for uri, r in self.extra_restrictions.items():
				gui_args.insert(0, r)
				gui_args.insert(0, uri)
				gui_args.insert(0, '--version-for')
		if self.source:
			gui_args.insert(0, '--source')
		if self.message:
			gui_args.insert(0, self.message)
			gui_args.insert(0, '--message')
		if self.cpu:
			gui_args.insert(0, self.cpu)
			gui_args.insert(0, '--cpu')
		if self.os:
			gui_args.insert(0, self.os)
			gui_args.insert(0, '--os')
		gui_args.append('--command')
		gui_args.append(self.command or '')

		return gui_args

	def get_extra_restrictions(self, iface_cache):
		"""Create list of L{model.Restriction}s for each interface, based on these requirements.
		@type iface_cache: L{zeroinstall.injector.iface_cache.IfaceCache}
		@rtype: tuple"""
		from zeroinstall.injector import model

		return dict((iface_cache.get_interface(uri), [model.VersionExpressionRestriction(expr)])
				for uri, expr in self.extra_restrictions.items())

	def _handle_restrictions(self, options):
		"""Gets the list of restrictions specified by the user.
		Handles --before, --not-before, --version and --version-for options.
		If a mapping isn't present, then the user didn't specify a restriction.
		If an entry maps to None, the user wishes to remove the restriction."""
		version = options.version

		interface_uri = self.interface_uri

		# Convert old --before and --not_before to new --version-for
		if options.before is not None or options.not_before is not None:
			if version is not None:
				raise SafeException("Can't use --before or --not-before with --version")
			if options.before or options.not_before:
				version = (options.not_before or '') + '..'
				if options.before:
					version += '!' + options.before
			else:
				version = ''		# Reset

		restrictions = dict((uri, (expr or None)) for (uri, expr) in (options.version_for or []))

		# Convert the --version short-cut to --version-for
		if version is not None:
			if interface_uri in restrictions:
				raise SafeException("Can't use --version and --version-for to set {uri}".format(uri = interface_uri))
			restrictions[interface_uri] = version or None

		return restrictions

File: ocaml/requirements.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** A structure representing constraints/requirements specified by the user *)

open General
open Support.Common

type requirements = {
  interface_uri : iface_uri;
  command : string option;
  source : bool;
  extra_restrictions : string StringMap.t;  (* iface -> range *)
  os : string option;
  cpu : string option;
  message : string option;
}

let default_requirements interface_uri = {
  interface_uri;
  command = Some "run";
  source = false;
  os = None;
  cpu = None;
  message = None;
  extra_restrictions = StringMap.empty;
}

let parse_extra json : string StringMap.t =
  let open Yojson.Basic.Util in
  let add map (key, value) =
    StringMap.add key (to_string value) map in
  List.fold_left add StringMap.empty (to_assoc json)

let parse_requirements json =
  let open Yojson.Basic.Util in
  let to_string_option_or_empty x =
    match to_string_option x with
    | Some "" -> None
    | v -> v in
  try
    let alist = to_assoc json in
    let r = ref (default_requirements "") in
    let before = ref None in
    let not_before = ref None in
    ListLabels.iter alist ~f:(fun (key, value) ->
      match key with
      | "before" -> before := to_string_option_or_empty value
      | "not_before" -> not_before := to_string_option_or_empty value
      | "interface_uri" -> r := {!r with interface_uri = to_string value}
      | "command" -> r := {!r with command = to_string_option value}
      | "source" -> r := {!r with source = to_bool value}
      | "extra_restrictions" -> r := {!r with extra_restrictions = (parse_extra value)}
      | "os" -> r := {!r with os = to_string_option value}
      | "cpu" ->  r := {!r with cpu = to_string_option value}
      | "message" -> r := {!r with message = to_string_option value}
      | _ -> raise_safe "Unknown requirements field '%s'" key
    );

    if !r.interface_uri = "" then
      raise_safe "Missing 'interface_uri'";

    (* Update old before/not-before values *)
    let () = 
      let use expr =
        assert (!r.extra_restrictions = StringMap.empty);
        r := {!r with extra_restrictions = StringMap.singleton !r.interface_uri expr} in
      match !not_before, !before with
      | None, None -> ()
      | Some low, None -> use low
      | None, Some high -> use @@ "..!" ^ high
      | Some low, Some high -> use @@ low ^ "..!" ^ high in

    !r
  with Type_error (msg, _) ->
    raise_safe "Bad requirements JSON: %s" msg

let load (system:system) path =
  let open Yojson.Basic in
  system#with_open_in [Open_rdonly; Open_binary] 0 path (fun ch ->
    try parse_requirements (from_channel ~fname:path ch)
    with Safe_exception _ as ex -> reraise_with_context ex "... parsing JSON file %s" path
  )

(** Updates requirements based on command-line options.
    Handles --before, --not-before, --version and --version-for options.
    Existing requirements can be removed by setting the version to the empty string (e.g. --version="").
    @return the options list with all version specifiers removed and the updated restrictions. *)
let parse_restrictions options default_iface extra_restrictions =
  let open Options in

  (* Perhaps we should initialise these from [extra_restrictions], but the Python version doesn't. *)
  let before = ref None in
  let not_before = ref None in

  (** Handle --before, --not-before and --version by converting to --version-for options *)
  let options = Support.Argparse.filter_map_options options (function
    | Before v -> before := Some v; None
    | NotBefore v -> not_before := Some v; None
    | RequireVersion v -> Some (RequireVersionFor (default_iface, v))
    | x -> Some x
  ) in

  let version = match !not_before, !before with
  | None, None -> None
  | Some low, None -> Some low
  | None, Some high -> Some ("..!" ^ high)
  | Some "", Some "" -> Some "" (* Reset *)
  | Some low, Some high -> Some (low ^ "..!" ^ high) in

  let options = match version with
  | None -> options
  | Some v -> ("--version", (RequireVersionFor (default_iface, v))) :: options in

  (** Process --version-for options *)
  let r = ref extra_restrictions in

  (** TODO: later options override earlier ones; issue an error instead *)
  let process = function
    | RequireVersionFor (iface, "") -> r := StringMap.remove iface !r; None    (* TODO: check no error if already removed *)
    | RequireVersionFor (iface, expr) -> r := StringMap.add iface expr !r; None
    | x -> Some x in
  let options = Support.Argparse.filter_map_options options process in
  (options, !r)

(** Update the requirements based on the options (used for e.g. "0install update APP"). *)
let parse_update_options ?(update=true) options requirements =
  let open Options in

  let (options, new_restrictions) = parse_restrictions options requirements.interface_uri requirements.extra_restrictions in

  let r = ref {requirements with extra_restrictions = new_restrictions} in

  let empty_to_opt = function
    | "" -> None
    | s -> Some s in

  let options = Support.Argparse.filter_map_options options (function
    | WithMessage v    -> r := {!r with message = empty_to_opt v}; None
    | Cpu v            -> r := {!r with cpu = empty_to_opt v}; None
    | Os v             -> r := {!r with os = empty_to_opt v}; None
    | SelectCommand v  -> r := {!r with command = empty_to_opt v}; None
    | Source when not update -> r := {!r with source = true}; None
    | Source when !r.source -> None
    | Source ->
          (** Partly because it doesn't make much sense, and partly because you
              can't undo it, as there's no --not-source option. *)
          raise_safe "Can't update from binary to source type!"
    | x -> Some x
  ) in

  (options, !r)

let parse_options options interface_uri ~command =
  parse_update_options ~update:false options @@ {(default_requirements interface_uri) with command}

(** Convert a set of requirements to the corresponding command-line options. *)
let to_options requirements =
  let opt_arg opt_name = function
    | None -> []
    | Some value -> [opt_name; value] in

  let version_for (iface, expr) = ["--version-for"; iface; expr] in

  List.concat @@ [
    if requirements.source then ["--source"] else [];
    opt_arg "--message" requirements.message;
    opt_arg "--cpu" requirements.cpu;
    opt_arg "--os" requirements.os;
    ["--command"; default "" (requirements.command)];
  ] @ List.map version_for @@ StringMap.bindings requirements.extra_restrictions
